src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -168,44 +261,55 @@ private SafeCurlHandle CreateRequestHandle(RequestCompletionSource state)
                 throw new HttpRequestException(SR.net_http_client_execution_error);
             }
 
-            Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_URL,
-                state.RequestMessage.RequestUri.OriginalString);
+            SetCurlOption(requestHandle, CURLoption.CURLOPT_URL, state.RequestMessage.RequestUri.AbsoluteUri);
             if (_automaticRedirection)
             {
-                Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_FOLLOWLOCATION, 1);
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_FOLLOWLOCATION, 1L);
+            }
+            if (state.RequestMessage.Content != null)
+            {
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_UPLOAD, 1L);
+            }
+            if (state.RequestMessage.Method == HttpMethod.Head)
+            {
+                SetCurlOption(requestHandle, CURLoption.CURLOPT_NOBODY, 1L);
             }
 
-            // TODO: Handle headers, proxy and other options
-
-            return requestHandle;
-        }
-
-        private HttpResponseMessage CreateResponseMessage(SafeCurlHandle requestHandle, HttpRequestMessage request)
-        {
-            var response = new HttpResponseMessage();
-            long httpStatusCode = 0;
-            int result = Interop.libcurl.curl_easy_getinfo(requestHandle, CURLINFO.CURLINFO_RESPONSE_CODE,
-                ref httpStatusCode);
-            if (CURLcode.CURLE_OK != result)
+            // Setup callbacks
+            SetCurlOption(requestHandle, CURLoption.CURLOPT_HEADERDATA, GCHandle.ToIntPtr(stateHandle));
+            SetCurlOption(requestHandle, CURLoption.CURLOPT_HEADERFUNCTION, (Interop.libcurl.curl_readwrite_callback)CurlReceiveHeadersCallback);
```

Comment:
This looks buggy.  You're allocating new delegates and marshaling them to the native function, but after this the delegates could get garbage collected.  You should make the callback static methods, create the delegates once and cache them in static readonly fields, and then pass those cached delegates in.  That'll not only help with such corruption, it'll also avoid allocating the delegates on each request.